pahfit/fitter.py

```diff
@@ -0,0 +1,209 @@
+    def register_starlight(self, name, temperature, tau):
+        """Register a starlight feature.
+
+        The exact representation depends on the implementation, but the
+        meaning of the parameters should be equivalent.
+
+        Parameters
+        ----------
+        name : str
+            Unique name. Will be used to allow retrieval of the results
+            after the fitting.
+
+        temperature : array of size 3
+            Blackbody temperature, given as [value, lower_bound,
+            upper_bound]. The bounds assume the same system as the
+            features table: if they are masked, the parameter will be
+            fixed, while +inf or -inf mean unbounded.
+
+        tau : array of size 3
+            Analogously, used as power.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_continuum(self, name, temperature, tau):
+        """Register a dust continuum feature."""
+        pass
+
+    @abstractmethod
+    def register_line(self, name, power, wavelength, fwhm):
+        """Register an emission line feature.
+
+        Typically a Gaussian profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_feature(self, name, power, wavelength, fwhm):
+        """Register a dust feature.
+
+        Typically a Drude profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_attenuation(self, name, tau):
+        """Register the S07 attenuation component.
+
+        Other types of attenuation might be possible in the future. Is
+        multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_absorption(self, name, tau, wavelength, fwhm):
+        """Register an absorption feature.
+
+        Typically a Drude profile. Is multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def evaluate_model(self, xz):
+        """Evaluate the model at the given wavelengths.
+
+        Parameters
+        ----------
+        xz : array
+            Rest frame wavelengths in micron
+
+        Returns
+        -------
+        yz : array
+            Rest frame flux in internal units
+
+        """
+        pass
+
+    @abstractmethod
+    def fit(self, xz, yz, uncz, maxiter=1000):
+        """Perform the fit using the framework of the subclass.
+
+        Fitter is unit agnostic, and deals with the numbers the Model
+        tells it to deal with. In practice, the input spectrum is
+        expected to be in internal units, and corrected for redshift
+        (models operate in the rest frame).
+
+        After the fit, the results can be retrieved via get_result().
+
+        Parameters
+        ----------
+        xz : array
+            Rest frame wavelengths in micron
+
+        yz : array
+            Rest frame flux in internal units.
+
+        uncz : array
+            Uncertainty on rest frame flux. Same units as yz.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_result(self, feature_name):
```

Comment:
Why wouldn't `fit` just return all of this info in one big table (or large numpy array with the right ordering)?  Rather than a `Features` table passing back and forth, we now just do a ton of method calls.  Not much of an improvement, IMO.  What about other "results" like covariance matrix, function evals, information on convergence, etc.?

Reply:
With the astropy framework, I can't "just" return a big table. I have to loop over the components and gather the results. 

So in my mind, every subclass have to implement some logic to create this table, and we would have to decide on a common format and guarantee that it is consistent between `Fitter` subclasses. (this might be another argument for just passing a Features table directly to the `Fitter`, and implementing the Features <-> fitting/results logic in the ABC).

As it currently stands `Fitter` only offers a way to retrieve a "single unit" of results. And the logic for how to map those units to the features table, is implemented in a single place: `Model._ingest_fit_result_to_features()`.

For the covariance matrix, we will have to come up with something else... but I feel it will still be possible to try and have a single place for the logic.

The fitting diagnostics / convergence information can be wildly different depending on the underlying fitting algorithm. I feel like understanding them full, will always require some knowledge of the specific framework. Though, if we can come up with a few "standard" diagnostics, then we can just provide access functions for those.

Reply:
I've definitely come around to your way of thinking on this since that old comment.  For "fit information", we can just ask the Fitter for some generic fit information for `Features.meta["fit_log"]` and insert it,  perhaps a dictionary, so each Fitter can define their own (with some attempt to normalize things e.g. `Features.meta["fit_log"]["function_evals"]` etc.). 

We can (later) ask the `Fitter` for the `covariance` too, and store it in `meta` as well.  We can plan to access that in a `Model.total_power` (i.e. total up various powers of the selected table, with uncertainty propagated through the full covariance matrix). And of course for `Features` pretty-printing.